crates/red_knot_python_semantic/src/semantic_index/use_def/bitset.rs

```diff
@@ -0,0 +1,402 @@
+        match self {
+            Self::Inline(blocks) => BitSetIterator::Inline {
+                blocks,
+                cur_block_index: 0,
+                cur_block: blocks[0],
+            },
+            Self::Heap(set) => BitSetIterator::Heap(set.iter()),
+        }
+    }
+}
+
+/// Iterator over values in a [`BitSet`].
+#[derive(Debug)]
+pub(super) enum BitSetIterator<'a, const B: usize> {
+    Inline {
+        /// The blocks we are iterating over.
+        blocks: &'a [u64; B],
+
+        /// The index of the block we are currently iterating through.
+        cur_block_index: usize,
+
+        /// The block we are currently iterating through (and zeroing as we go.)
+        cur_block: u64,
+    },
+    Heap(btree_set::Iter<'a, u32>),
+}
+
+impl<const B: usize> Iterator for BitSetIterator<'_, B> {
+    type Item = u32;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Self::Inline {
+                blocks,
+                cur_block_index,
+                cur_block,
+            } => {
+                while *cur_block == 0 {
+                    if *cur_block_index == B - 1 {
+                        return None;
+                    }
+                    *cur_block_index += 1;
+                    *cur_block = blocks[*cur_block_index];
+                }
+                let value = cur_block.trailing_zeros();
+                // reset the lowest set bit
+                *cur_block &= cur_block.wrapping_sub(1);
+                // SAFETY: `value` cannot be more than 64, `cur_block_index` cannot be more than
+                // `B - 1`, and we check above that `B * 64 < u32::MAX`. So both `64 *
+                // cur_block_index` and the final value here must fit in u32.
+                #[allow(clippy::cast_possible_truncation)]
+                Some(value + (64 * *cur_block_index) as u32)
+            }
+            Self::Heap(set_iter) => set_iter.next().copied(),
+        }
+    }
+}
+
+impl<const B: usize> std::iter::FusedIterator for BitSetIterator<'_, B> {}
+
+/// Array of [`BitSet<B>`]. Up to N stored inline, more than that overflows to a Vector.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(super) enum BitSetArray<const B: usize, const N: usize> {
+    Inline {
+        /// Array of N [`BitSet`].
+        array: [BitSet<B>; N],
+
+        /// How many of the bitsets are used?
+        size: usize,
+    },
+
+    Heap(Vec<BitSet<B>>),
+}
+
+impl<const B: usize, const N: usize> Default for BitSetArray<B, N> {
+    fn default() -> Self {
+        Self::Inline {
+            array: std::array::from_fn(|_| BitSet::default()),
+            size: 0,
+        }
+    }
+}
+
+impl<const B: usize, const N: usize> BitSetArray<B, N> {
+    /// Create a [`BitSetArray`] of `size` empty [`BitSet`]s.
+    pub(super) fn of_size(size: usize) -> Self {
+        let mut array = Self::default();
+        for _ in 0..size {
+            array.push(BitSet::default());
+        }
+        array
```

Comment:
I'm not sure if the compiler can see through this. Might be worth to have two extra branches
```suggestion
				if size <= N {
					Self::default()
				} else {
					Self::Heap(Vec::with_capacity(size))
				}
```

Reply:
This suggestion won't work for the reason described above: the semantics of this method are not `with_capacity` semantics. The first branch would also need to increment `self.size`, and the second branch needs to also actually push `size` elements onto the vec, after creating it.

But it still may be worth the separate branches here.

Reply:
The two explicit branches don't seem to affect benchmark performance detectably, but the code is still quite concise and I think it is likely the compiler doesn't see through the other version, so I'll leave the two branches.